File: 0022-generate-parentheses/0022-generate-parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        result = []
        self.generate(result, "", 0, 0, n)
        return result
    
    def generate(self, result: List[str], current: str, open: int, close: int, max: int):
        if len(current) == max * 2:
            result.append(current)
            return
        
        if open < max:
            self.generate(result, current + "(", open + 1, close, max)
        
        if close < open:
            self.generate(result, current + ")", open, close + 1, max)

File: 0022-generate-parentheses/0022-generate-parentheses.java
class Solution {
    public List<String> generateParenthesis(int n) {
        List <String> resu = new ArrayList<>();
        backtrack(resu,"",0,0,n);
        return resu;
    }

    private void backtrack(List<String> resu,String current, int open, int close, int n){
        if(current.length() == n*2){
            resu.add(current);
            return;
        }

        if(open < n){
            backtrack(resu,current + "(" , open+1, close, n);
        }
        if(close < open){
            backtrack(resu, current + ")", open, close+1, n);
        }
    }
}